src/Data/Tree/DUAL/Internal.hs

```diff
@@ -19,107 +22,77 @@
+-- "Data.Tree.DUAL" is one extra type used in the implementation of
+-- 'DUALTree', along with functions for manipulating them. A type of
+-- /non-empty/ trees without up annotations, 'DALTree', is defined. A
+-- 'DUALTree' is a 'DALTree' with a top-level @u@ annotation along with
+-- a possible 'EmptyDUAL'. This method has the advantage that the type
+-- system enforces the invariant that there is only one representation
+-- for the empty tree. It also allows us to get away with only
+-- 'Semigroup' constraints in many places.
 --
 -----------------------------------------------------------------------------
 
 module Data.Tree.DUAL.Internal
-       (
-         -- * DUAL-trees
-         DUALTreeNE(..), DUALTreeU(..), DUALTree(..)
-
-         -- * Constructing DUAL-trees
-       , empty, leaf, leafU, annot, applyD
-
-         -- * Modifying DUAL-trees
-       , applyUpre, applyUpost
-       , mapUNE, mapUU, mapU
-
-         -- * Accessors and eliminators
-       , nonEmpty, getU, foldDUALNE, foldDUAL, flatten
-
-       ) where
-
-import           Control.Arrow      ((***))
-import           Data.List.NonEmpty (NonEmpty (..))
-import qualified Data.List.NonEmpty as NEL
-import           Data.Maybe         (fromMaybe)
+  (
+    -- * DUAL-trees
+    DALTree(..), DUALTree(..)
+
+    -- * Constructing DUAL-trees
+  , leaf, leafU, down, annot
+
+    -- * Folding DUAL-trees
+  , foldDUAL
+  , foldDUAL'
+  , flatten
+
+    -- * Up annotations
+  , _u
+  , getU
+  , mapU
+  , preapplyU
+  , postapplyU
+
+  ) where
+
+#if __GLASGOW_HASKELL__ < 710
+import           Control.Applicative
+#endif
+import           Control.DeepSeq
+import           Data.Foldable             as F (foldMap)
 import           Data.Monoid.Action
+import           Data.Monoid.WithSemigroup
 import           Data.Semigroup
+import           Data.Sequence
 import           Data.Typeable
 
-import           Control.Newtype
+------------------------------------------------------------------------
+-- DALTree
+------------------------------------------------------------------------
 
-------------------------------------------------------------
--- DUALTreeNE
-------------------------------------------------------------
-
--- | /Non-empty/ DUAL-trees.
-data DUALTreeNE d u a l
-  = Leaf   u l        -- ^ Leaf with data value and @u@ annotation
-  | LeafU  u          -- ^ Leaf with only @u@ annotation
-  | Concat (NonEmpty (DUALTreeU d u a l))
-                      -- ^ n-way branch, containing a /non-empty/ list
-                      --   of subtrees.
-  | Act    d (DUALTreeU d u a l)
-                      -- ^ @d@ annotation
-  | Annot  a (DUALTreeU d u a l)
-                      -- ^ Internal data value
+-- | /Non-empty/ DUAL-tree without the u.
+data DALTree d a l
+  = Leaf   !l                  -- ^ @l@eaf
+  | Down   !d !(DALTree d a l) -- ^ @d@own-annotation
+  | Annot  !a !(DALTree d a l) -- ^ @a@nnotation
+  | Concat (Seq (DALTree d a l)) -- ^ n-way branch
```

Comment:
Never mind, I see later that we do use `Concat mempty`, in the implemenation of `leafU`.  But that means describing a `DALTree` as non-empty is a bit misleading.